### PR TITLE
Inefficient permission check in InstructorFeedbackSessionActions #4506

### LINKED ISSUE #4506: Inefficient permission check in InstructorFeedbackSessionActions
If an instructor has no permissions to submit responses for any sections, he cannot click the submit button. This logic is wrong as he might still have questions not directed to any student.

Performing this check also slows down performance by alot.

```
for (String section : sectionsInCourse) {
       if (instructor.isAllowedForPrivilege(section, 
                                                 session.feedbackSessionName, 
                                                 Const.ParamsNames.INSTRUCTOR_PERMISSION_SUBMIT_SESSION_IN_SECTIONS)) {
                shouldEnableSubmitLink = true;
                break;
            }
        }
```

### PR DESCRIPTION
**Pull request by LowWeiLin**
Fixes #4506

### CODE REVIEW

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java [hunk 1]
```diff
@@ -344,6 +344,10 @@ public boolean isAllowedForPrivilege(String sectionName, String sessionName, Str
         return isAllowedInSessionLevel(sectionName, sessionName, privilegeName);
     }
     
+    public boolean isAllowedForPrivilegeAnySection(String sessionName, String privilegeName) {
```
Comment (damithc) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Public method. Add comment?

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java [hunk 2]
```diff
@@ -434,6 +441,22 @@ private boolean isAllowedInSessionLevel(String sectionName, String sessionName,
         }
     }
     
+    private boolean isAllowedInSessionLevelAnySection(String sessionName, String privilegeName) {
+        for (String sectionName : this.sessionLevel.keySet()) {
+            if (this.sessionLevel.get(sectionName).get(sessionName).containsKey(privilegeName) &&
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Should there be a check for <code>this.sessionLevel.get(sectionName).containsKey(sessionName)</code> ?

Comment (LowWeiLin) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Yes? Otherwise the next condition might give a null.

Comment (LowWeiLin) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Oh i get what you mean now. No need to check for it as we are iterating through the key set. It must exist.

Comment (LowWeiLin) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Updated/refactored.

File: src/main/java/teammates/common/datatransfer/InstructorPrivileges.java [hunk 3]
```diff
@@ -434,6 +441,22 @@ private boolean isAllowedInSessionLevel(String sectionName, String sessionName,
...
     
+    private boolean isAllowedInSessionLevelAnySection(String sessionName, String privilegeName) {
+        HashSet<String> sections = new HashSet<String>(this.sessionLevel.keySet());
+        sections.addAll(this.sectionLevel.keySet());
+        for (String sectionName : sections) {
+            if (isAllowedInSessionLevel(sectionName, sessionName, privilegeName)) {
+                return true;
+            }
+        }
+        for (String sectionName : this.sectionLevel.keySet()) {
```
Comment (kanghj) [inline src/main/java/teammates/common/datatransfer/InstructorPrivileges.java]: Is this for loop needed?

**Comment by kanghj** [comment]
Looks good to merge. :+1:

**Pull request merged**